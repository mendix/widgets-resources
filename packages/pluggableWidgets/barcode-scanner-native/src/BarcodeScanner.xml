<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.native.barcodescanner.BarcodeScanner" supportedPlatform="Native" needsEntityContext="true" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Barcode scanner</name>
    <description>Scan barcode and QR code values.</description>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAMAAACdt4HsAAAAElBMVEUAAABXV1dVVVVJsPhIsPdVVVWcEer3AAAABHRSTlMASeP9eof+WwAAAHJJREFUeAHtlzEKxTAMQ239r/tfuaHtFEpxIkog6C1e7AchixQXJsET5OQ2eIMo8LDNOQE7wegTOkEMspsAJGKE/iaBjDrlG7OAf6M89xdYYIEFBb4XtPF7n/sL1v2CWYoespSY56ysFw698uilS6995gADwwwpR7MpFwAAAABJRU5ErkJggg==</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="barcode" type="attribute">
                    <caption>Barcode</caption>
                    <description>The attribute that will receive the scanned barcode value.</description>
                    <attributeTypes>
                        <attributeType name="String"/>
                    </attributeTypes>
                </property>
            </propertyGroup>
            <propertyGroup caption="Mask">
                <property key="showMask" type="boolean" required="true" defaultValue="false">
                    <caption>Show barcode mask</caption>
                    <description />
                </property>
                <property key="showAnimatedLine" type="boolean" required="true" defaultValue="false">
                    <caption>Show animated line</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onDetect" type="action" required="false">
                    <caption>On detect</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Name"/>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
