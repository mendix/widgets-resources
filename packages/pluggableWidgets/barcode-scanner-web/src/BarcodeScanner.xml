<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.barcodescanner.BarcodeScanner" offlineCapable="true"
    xmlns="http://www.mendix.com/widget/1.0/"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd" pluginWidget="true">
    <name>Barcode Scanner</name>
    <description>The widget lets you scan a barcode</description>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAFgSURBVHgB7duxboMwEAbgP1GGCmZ4HtZOXYCZR4MVpk5deR9mULe2DKlUgrnaItJZ939TAibCJ3OWyfny9QOGXWEcAwDjbtsD4ziibVssy7J7QZZlqKoKRVFAk9D7fhgBwzA4f2Q1TRO6roM2ofd93WsoSdMU2szzHNTGOwesnS/LEto0TYM8z+HrJjXo+x4xWJ/tvbxU1/XhdZwGYRwDsD2QJMnv55Ckoo3Un4cA3LOp1mzvS+rPhatB4xgAGMcAwDhxLXD3+v755/vH20tU5134CMA4BgDGMQAwjgGAcQwAjGMAYBwDAOP4PgAnkdbfzz4fiiMAJ9H6xkfCJAjjmANwEs4CkeK/wzCOAYBxDACMc1aLr0VFGqvCfUn9cVaLa60K9yX157Ba/D8V2NpJ/WEShHGcBaQG23r7WPcMuXiPgFj3DLkcVou7xLpnyKta3CXWPUOsFnfgNAjjzAfgG6Bat42JKk75AAAAAElFTkSuQmCC</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <!-- TODO: required should be true -->
                <property key="datasource" type="attribute" required="false">
                    <caption>Barcode</caption>
                    <description>The attribute that will receive the scanned barcode value.</description>
                    <attributeTypes>
                        <attributeType name="String"/>
                    </attributeTypes>
                </property>
            </propertyGroup>

            <propertyGroup caption="Mask">
                <property key="showMask" type="boolean" required="true" defaultValue="false">
                    <caption>Show barcode mask</caption>
                    <description />
                </property>
            </propertyGroup>

            <propertyGroup caption="Events">
                <property key="onDetect" type="action" required="false">
                    <caption>On detect</caption>
                    <description />
                </property>
                <property key="onClose" type="action" required="false">
                    <caption>On close</caption>
                    <description />
                </property>
            </propertyGroup>

            <propertyGroup caption="Common">
                <systemProperty key="Name"/>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
