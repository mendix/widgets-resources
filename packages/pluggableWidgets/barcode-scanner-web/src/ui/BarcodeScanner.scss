.widget-barcode-scanner-container {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background: #000000;
    // Should be higher than the z-index of `.layout-atlas .region-sidebar` so it sits on top of the page.
    z-index: 110;

    .video {
        position: relative;
        width: 100%;
        height: 100%;
    }

    .close-button {
        position: absolute;
        top: 30px;
        left: 30px;
        color: white;
        .glyphicon {
            font-size: 16px;
        }
    }

    .video-canvas {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
        display: flex;
        flex-direction: row;
        align-items: stretch;

        .canvas-background {
            background-color: rgba(0, 0, 0, 0.7);
        }

        .canvas-left {
            flex: 1;
        }

        .canvas-middle {
            flex: 0;
            flex-basis: 60%;
            display: flex;
            flex-direction: column;

            .canvas-middle-top {
                flex: 1;
            }

            .canvas-middle-middle {
                flex: 0;
                padding-top: 100%;
                position: relative;

                $corner-offset: 13px;
                .corner {
                    position: absolute;
                    width: 32px;
                    height: 32px;
                    box-sizing: border-box;
                    $corner-border: 5px solid #ffffff;
                    border-left: $corner-border;
                    border-top: $corner-border;

                    &.corner-top-left {
                        top: -$corner-offset;
                        left: -$corner-offset;
                    }
                    &.corner-top-right {
                        top: -$corner-offset;
                        right: -$corner-offset;
                        transform: rotate(90deg);
                    }
                    &.corner-bottom-right {
                        bottom: -$corner-offset;
                        right: -$corner-offset;
                        transform: rotate(180deg);
                    }
                    &.corner-bottom-left {
                        bottom: -$corner-offset;
                        left: -$corner-offset;
                        transform: rotate(270deg);
                    }
                }
            }

            .canvas-middle-bottom {
                flex: 1;
            }
        }

        .canvas-right {
            flex: 1;
        }
    }
}
