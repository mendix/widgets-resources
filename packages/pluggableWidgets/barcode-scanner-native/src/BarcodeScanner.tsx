import { flattenStyles } from "@native-mobile-resources/util-widgets";
import { ValueStatus } from "mendix";
import { Component, createElement } from "react";
import { View } from "react-native";
import { RNCamera } from "react-native-camera";

import { BarcodeScannerProps } from "../typings/BarcodeScannerProps";
import { BarcodeScannerStyle, defaultBarcodeScannerStyle } from "./ui/styles";
import { executeAction } from "@widgets-resources/piw-utils";

export type Props = BarcodeScannerProps<BarcodeScannerStyle>;

export class BarcodeScanner extends Component<Props> {
    private readonly styles = flattenStyles(defaultBarcodeScannerStyle, this.props.style);
    private readonly onBarCodeReadHandler = this.onBarCodeRead.bind(this);

    render(): JSX.Element {
        return (
            <View style={this.styles.container}>
                <RNCamera
                    testID={this.props.name}
                    style={{ flex: 1, justifyContent: "center", alignItems: "center" }}
                    captureAudio={false}
                    onBarCodeRead={this.onBarCodeReadHandler}
                />
            </View>
        );
    }

    private onBarCodeRead(event: { data: string }): void {
        if (this.props.barcode.status !== ValueStatus.Available || event.data === this.props.barcode.value) {
            return;
        }

        this.props.barcode.setValue(event.data);

        executeAction(this.props.onDetect);
    }
}
