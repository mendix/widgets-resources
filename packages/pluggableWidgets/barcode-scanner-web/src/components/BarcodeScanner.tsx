import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import classNames from "classnames";
import { useCodeScanner } from "../hooks/useCodeScanner";
import { browserSupportsCameraAccess, useMediaStream } from "../hooks/useMediaStream";

import "../ui/BarcodeScanner.scss";

export interface BarcodeScannerProps {
    onClose?: () => void;
    onDetect?: (data: string) => void;
    showMask: boolean;
}

export function BarcodeScanner({ onClose, onDetect, showMask }: BarcodeScannerProps): ReactElement | null {
    const [showScannerOverlay, setShowScannerOverlay] = useState<boolean>(true);
    const [videoElement, setVideoElement] = useState<HTMLVideoElement | null>(null);
    const hasDetectedOnce = useRef<boolean>(false);
    const { streamObject, cleanupStreamObject, error } = useMediaStream();
    const { codeResult } = useCodeScanner(streamObject, videoElement);
    const supportsCameraAccess = browserSupportsCameraAccess();

    const updateVideoElement = useCallback(
        (node: HTMLVideoElement | null) => {
            setVideoElement(node);
        },
        [setVideoElement]
    );

    const play = useCallback((event: React.SyntheticEvent<HTMLVideoElement, Event>) => {
        if (event.currentTarget.paused) {
            event.currentTarget.play(); // TODO: doesn't work on iOS safari
        }
    }, []);

    function onCloseOverlay(): void {
        setShowScannerOverlay(false);
        cleanupStreamObject();
    }

    useEffect(() => {
        if (videoElement && streamObject) {
            videoElement.srcObject = streamObject;
        }
    }, [streamObject, videoElement]);

    useEffect(() => {
        if (onDetect && codeResult && !hasDetectedOnce.current) {
            hasDetectedOnce.current = true;
            onDetect(codeResult);
        }
    }, [codeResult, onDetect]);

    if (!showScannerOverlay) {
        return null;
    }
    if (!supportsCameraAccess) {
        // Mendix ensures that Mendix apps are only run in the supported browsers and all of them
        // support the `navigator.mediaDevices.getUserMedia` API. So no additional error handling
        // needs to be done, but just in case we soft catch it.
        return <div>This browser is not compatible with the barcode scanner widget</div>; // TODO: improve message, since this will also show when no https is used in certain browsers
    }
    if (error && error !== "ERROR_NOT_ALLOWED") {
        return <div>{error}</div>;
    }
    return (
        <div className={classNames("widget-barcode-scanner-container")}>
            <video className={classNames("video")} ref={updateVideoElement} onCanPlay={play} />
            {showMask ? (
                <div className={classNames("video-canvas")}>
                    <div className={classNames("canvas-left", "canvas-background")} />
                    <div className={classNames("canvas-middle")}>
                        <div className={classNames("canvas-middle-top", "canvas-background")} />
                        <div className={classNames("canvas-middle-middle")}>
                            <div className={classNames("corner", "corner-top-left")} />
                            <div className={classNames("corner", "corner-top-right")} />
                            <div className={classNames("corner", "corner-bottom-right")} />
                            <div className={classNames("corner", "corner-bottom-left")} />
                        </div>
                        <div className={classNames("canvas-middle-bottom", "canvas-background")} />
                    </div>
                    <div className={classNames("canvas-right", "canvas-background")} />
                </div>
            ) : null}
            <button className={classNames("btn btn-image btn-icon close-button")} onClick={onClose || onCloseOverlay}>
                <div className={classNames("glyphicon", "glyphicon-remove")} />
            </button>
        </div>
    );
}
