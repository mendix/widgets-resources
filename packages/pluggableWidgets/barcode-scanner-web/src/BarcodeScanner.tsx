import { createElement, FunctionComponent, useCallback } from "react";
import { BarcodeScannerContainerProps } from "../typings/BarcodeScannerProps";
import { BarcodeScanner as BarcodeScannerComponent } from "./components/BarcodeScanner";

export const BarcodeScanner: FunctionComponent<BarcodeScannerContainerProps> = props => {
    const onClose = useCallback(() => props.onClose?.execute(), [props.onClose]);
    return <BarcodeScannerComponent onClose={props.onClose ? onClose : undefined} />;
};
