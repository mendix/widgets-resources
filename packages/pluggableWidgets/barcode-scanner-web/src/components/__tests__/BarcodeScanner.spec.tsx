import { mount, shallow } from "enzyme";
import { createElement } from "react";
import { BarcodeScanner } from "../BarcodeScanner";

describe("Barcode scanner", () => {
    const backupMediaDevices = window.navigator.mediaDevices;
    afterEach(() => {
        // reset the mocking
        Object.defineProperty(window.navigator, "mediaDevices", {
            value: backupMediaDevices,
            writable: true
        });
    });
    function mockGetUserMedia(getUserMediaMock: jest.Mock): void {
        Object.defineProperty(window.navigator, "mediaDevices", {
            value: {
                getUserMedia: getUserMediaMock
            },
            writable: true
        });
    }

    it("shows an appropriate error when the mediaDevices API is not present (like over http)", () => {
        expect(navigator.mediaDevices).toBe(undefined);
        const barcodeScanner = mount(<BarcodeScanner showMask />);
        expect(barcodeScanner.text()).toBe("This browser is not compatible with the barcode scanner widget");
    });

    it("shows a loading screen while waiting for the user to give persmission to access the media device", () => {
        mockGetUserMedia(jest.fn());
        expect(shallow(<BarcodeScanner showMask />)).toMatchSnapshot();
    });

    it("does not show the overlay when the user opts out of it", () => {
        mockGetUserMedia(jest.fn());
        expect(shallow(<BarcodeScanner showMask={false} />)).toMatchSnapshot();
    });

    it("shows the appropriate error for the user", () => {
        mockGetUserMedia(
            jest.fn(() => {
                throw new Error("This is the error message");
            })
        );

        const barcodeScanner = mount(<BarcodeScanner showMask />);
        expect(barcodeScanner.text()).toBe("This is the error message");
    });
});
