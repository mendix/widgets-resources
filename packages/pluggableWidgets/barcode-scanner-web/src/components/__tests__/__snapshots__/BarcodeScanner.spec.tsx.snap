// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Barcode scanner does not show the overlay when the user opts out of it 1`] = `
<div
  className="widget-barcode-scanner-container"
>
  <video
    className="video"
    onCanPlay={[Function]}
  />
  <button
    className="btn btn-image btn-icon close-button"
    onClick={[Function]}
  >
    <div
      className="glyphicon glyphicon-remove"
    />
  </button>
</div>
`;

exports[`Barcode scanner shows a loading screen while waiting for the user to give persmission to access the media device 1`] = `
<div
  className="widget-barcode-scanner-container"
>
  <video
    className="video"
    onCanPlay={[Function]}
  />
  <div
    className="video-canvas"
  >
    <div
      className="canvas-left canvas-background"
    />
    <div
      className="canvas-middle"
    >
      <div
        className="canvas-middle-top canvas-background"
      />
      <div
        className="canvas-middle-middle"
      >
        <div
          className="corner corner-top-left"
        />
        <div
          className="corner corner-top-right"
        />
        <div
          className="corner corner-bottom-right"
        />
        <div
          className="corner corner-bottom-left"
        />
      </div>
      <div
        className="canvas-middle-bottom canvas-background"
      />
    </div>
    <div
      className="canvas-right canvas-background"
    />
  </div>
  <button
    className="btn btn-image btn-icon close-button"
    onClick={[Function]}
  >
    <div
      className="glyphicon glyphicon-remove"
    />
  </button>
</div>
`;
